### PR TITLE
Implement `Rename` on top of `MarkOccurences`

### PR DESCRIPTION
**Pull request by mlangc**
Consult [#1002698](https://app.assembla.com/spaces/scala-ide/tickets/1002698-markoccurences-and-rename-refactoring-should-share-common-implementation-for--quot-find-all-occurences-quot-/details#) as well as the commit messages for further background.

### CODE REVIEW

**Comment by mlangc** [comment]
Please do not merge, as I'm investigating a potential issue.

**Comment by wpopielarski** [comment]
not sure if it is your issue. I just committed a change which broke jenkins
build. Try to build it again from your branch. Code inspection on Monday
from my side.

2016-06-18 13:33 GMT+02:00 Matthias Langer <notifications@github.com>:

**Comment by mlangc** [comment]
@wpopielarski no, the problem I was looking into was definitely home made :-). Anyway, it should be fixed now.

File: src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala [hunk 1]
```diff
@@ -5,67 +5,229 @@
...
+     * Unfortunately, the Scala compiler does not set modifier positions for the DefDef, which is a problem for us,
+     * because we are relying on the DefDef when printing out source code. The following function patches the affected
+     * DefDefs accordingly.
+     *
+     * See #1002392 and #1002502
+     */
+    def eventuallyFixModifierPositionsForLazyVals(t: Tree): Unit = t match {
+      case dd: DefDef if dd.mods.isLazy && dd.mods.positions.isEmpty =>
+        val vd = selection.root.find {
+          case vd: ValDef if vd.mods.isLazy && !vd.mods.positions.isEmpty && dd.pos.point == vd.pos.point => true
```
Comment (wpopielarski) [inline src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala]: just curios what 'point' in position is used for

File: src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala [hunk 2]
```diff
@@ -5,67 +5,229 @@
...
-    val selectedTree = (new FileSelection(file, global.unitOfFile(file).body, from, to)).findSelectedWithPredicate {
-      case (_: global.TypeTree) | (_: global.SymTree) | (_: global.Ident) => true
-      case _ => false
+    /*
+     * ImportSelectorTrees need special treatment, since it is not a priory clear which side (the name or rename side) of
+     * the tree is relevant for us.
+     */
+    def positionsWithPotentiallyReferencingTrees(trees: Seq[Tree]): Seq[(Position, Tree)] = {
+      trees.flatMap {
+        case ImportSelectorTree(name, rename) => Seq(name.namePosition() -> name, rename.namePosition() -> rename)
```
Comment (wpopielarski) [inline src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala]: just a question, is it better to work with `ImportSelectorTrees` or pure `Imports`? Honestly I don't work with enriched trees and looks that I'm a bit reckless.

File: src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala [hunk 3]
```diff
@@ -5,67 +5,229 @@
...
+            if (consumedId == selection.name) {
+              Some((new RangePosition(np.source, markerAtSymStart, markerAtSymStart, markerAtSymStart + selection.name.length), occ))
+            } else {
+              trace(s"consumedId `$consumedId` does not match selecton name `${selection.name}`")
+              None
+            }
+          }
+
+        case op: OffsetPosition =>
+          // Normally, we would not like to deal with offset positions here at all.
```
Comment (wpopielarski) [inline src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala]: cool!

File: src/test/scala/scala/tools/refactoring/tests/implementations/MarkOccurrencesTest.scala [hunk 4]
```diff
@@ -439,4 +438,74 @@ class MarkOccurrencesTest extends TestHelper {
       def doit = for (#(x) <- Seq(#(1), #(2))) yield x
     }
     """)
+
+  @Test
+  def trivialObject() = markOccurrences("""
+    object DasDing/*<-cursor-3*/
```
Comment (wpopielarski) [inline src/test/scala/scala/tools/refactoring/tests/implementations/MarkOccurrencesTest.scala]: considering German words in code :), in Polish we have 'wi-hajst-er' (written together, hyphens for readability) for Ding

**Comment by wpopielarski** [comment]
Great improvement, I have to start using `Movements` and `SourceWithMarker` definitely. LGTM

File: src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala (hunk 1 above)
Comment (mlangc) [inline src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala]: To point is used to make sure that we are indeed selecting the right `ValDef` to patch.

File: src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala [hunk 5]
```diff
@@ -5,67 +5,229 @@
...
+          case _ => false
+        }
+
+        vd.foreach { vd =>
+          // Note that we patch the DefDef in place because we need the correctly set positions in the tree
+          // later in the refactoring process. A precursor of this code pretended to be pure, by copying
+          // dd and then calling 'mods.setPositions' on the copy. This had exactly the same (needed) side effects
+          // however, as the Modifier object affected by setPositions was the same anyway. If you are interested
+          // in the related discussion, take a look at https://github.com/scala-ide/scala-refactoring/pull/78.
+          dd.mods.setPositions(vd.asInstanceOf[ValDef].mods.positions)
```
Comment (mlangc) [inline src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala]: I could indeed avoid a cast by using `collect`, but then I would continue traversing the tree after a result has been found. What I really would like to use is `collectFirst`, which, unfortunately, no one cared to implement so far.

File: src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala (hunk 2 above)
Comment (mlangc) [inline src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala]: Hmm, I can't answer that in general.. in my case using `ImportSelectorTrees` was just the most strait forward solution.

File: src/test/scala/scala/tools/refactoring/tests/implementations/MarkOccurrencesTest.scala (hunk 4 above)
Comment (mlangc) [inline src/test/scala/scala/tools/refactoring/tests/implementations/MarkOccurrencesTest.scala]: :-) feel free to add some Polish flavor to our unit tests too.

File: src/test/scala/scala/tools/refactoring/tests/implementations/MarkOccurrencesTest.scala [hunk 6]
```diff
@@ -439,4 +438,74 @@ class MarkOccurrencesTest extends TestHelper {
...
+   @Test
+   def onConsOperator() = markOccurrences("""
+     object Constanzia {
+       val a = 1 :: 2 :: 3 :: Nil
+       val b = '1' :: '2' :: Nil
+       val c = 1.1 :: 2.2 ::/*<-cursor*/ Nil
+     }
+   """, """
+     object Constanzia {
+       val a = 1 ## 2 ## 3 ## Nil
```
Comment (sschaef) [inline src/test/scala/scala/tools/refactoring/tests/implementations/MarkOccurrencesTest.scala]: This still doesn't work for me in the IDE. Does it work for you?

File: src/test/scala/scala/tools/refactoring/tests/util/TextSelections.scala [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+import scala.collection.immutable.Queue
+import scala.annotation.tailrec
+
+/**
+ * Helper class to extract selections from unit test input.
+ *
+ * The following markers are supported:
+ *
+ * <ul>
```
Comment (sschaef) [inline src/test/scala/scala/tools/refactoring/tests/util/TextSelections.scala]: Scaladoc supports markdown like list syle.

File: src/test/scala/scala/tools/refactoring/tests/implementations/MarkOccurrencesTest.scala (hunk 6 above)
Comment (mlangc) [inline src/test/scala/scala/tools/refactoring/tests/implementations/MarkOccurrencesTest.scala]: No, it doesn't, although the unit tests prove that the problem lies not within the refactoring library (at least not any more). I'll file an IDE ticket for this as soon as this PR is merged.

File: src/test/scala/scala/tools/refactoring/tests/util/TextSelections.scala (hunk 7 above)
Comment (mlangc) [inline src/test/scala/scala/tools/refactoring/tests/util/TextSelections.scala]: Hmm, you mean something like
```scala
 /**
   * Title
   *
   * * Item 1
   * * Item 2
   */
```

At least in the ScalaIDE hovers, this does not work for me....

Comment (sschaef) [inline src/test/scala/scala/tools/refactoring/tests/util/TextSelections.scala]: I think you need to use a dash: https://wiki.scala-lang.org/display/SW/Syntax

**Comment by sschaef** [comment]
Alright, LGTM then. Would be very welcome if you also fix the integration into the IDE.

**Comment by mlangc** [comment]
@sschaef I'll look into it (see [ticket 1002703](https://app.assembla.com/spaces/scala-ide/tickets/1002703-markoccurrences-doesn--39-t-work-with-operators/details#) and [ticket 1002703](https://app.assembla.com/spaces/scala-ide/tickets/1002703-markoccurrences-doesn--39-t-work-with-operators/details#))

**Pull request merged**